core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -2455,7 +2455,13 @@ protected void convertCollectionTable(
     }
 
     RexNode rexCall = bb.convertExpression(call);
-    final List<RelNode> inputs = bb.retrieveCursors();
+    List<RelNode> inputs = new ArrayList<>();
+
+    if (rexCall.getKind() == SqlKind.TUMBLE) {
```

Comment:
I found I need this line to make `TABLE` parameter supported. Otherwise the inputs to TUMBLE table function will be empty even though there is a `TABLE` parameter and parser successfully parses it to a `SqlSelect`.

Reply:
It's possible that the fix is in `bb.retrieveCursors()`. Also possible, and in fact better, would be to let Calcite treat `TUMBLE` as an ordinary table function (in other words, don't check for `SqlKind.TUMBLE`). Then everything might just work out of the box.

Reply:
I removed this check here and give it a try: confirmed again Inputs will be empty (e.g. TABLE parameter is not converted). Am I doing right to let treat `TUMBLE` as an ordinary table function? Is it actually other place to not check TUMBLE to make it as a ordinary table function?

Reply:
@julianhyde do you have more comments regarding to this line of change?

Also generally I will try to push this change into 1.22.0 release. Could you let me know if there are other big refactors that I can do so I can do them soon?

Or if you think it is less mature to let this change into 1.22.0, please let me know and it's ok. I will work toward to make it released in 1.23.0

Reply:
In order to hack into this method, we may match the `TUMBLE` table value function just like lateral table [1]

[1] https://github.com/apache/calcite/blob/b93ec5a9edb7459696385e6adad67b92b6d406d7/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java#L2199

Reply:
I changed the `SqlKind` of TUMBLE table-valued funciton to `COLLECTION_TABLE`. It will have the same problem that the input of `TableFunctionScan` is empty, so basically it does not fix the `bb.retrieveCursors();`